### PR TITLE
Jetpack: Add/module registration refactor

### PR DESCRIPTION
**Pull request by enejb**
#### Changes proposed in this Pull Request:
Start of refactoring out modules code out of the main Jetpack class into Module Registration class and a Module class. 

This is an attempt to make more sense of the jetpack class by splitting it up some of it up into a module registry that would be responsible for all the different module retrial as well as activation etc.. functionality that is currently available jetpack class.

This PR will not try to do everything. Just the first step. Which is implementing a way to register a module and then a simple way to retrieve the module.  That is compatible with the current Jetpack class.

Currently use on the header comment in the {module-slug}.php files. Which gets cached in transients. 
This PR proposed a tool to build the registration functionality and store the module information in `code/run time` using the same header comments. 

We do this by first running ` php tools/build-module-registration.php` from the jetpack directly. This  creates a new `modules/module-registration.php` file. That has all the calls to `jetpack_register_module()`

Which servers as the way to `register` the modules in jetpack. 
This `jetpack_register_module` works in very similar way as the cores `register_block_type` function. 

[ ] write some more tests. 

#### Testing instructions:
Do the new tests pass? 

#### Proposed changelog entry for your changes:
* Add a way to register modules.

### CODE REVIEW

**Review by kraftbj — COMMENTED** [review]
I like this approach and that it auto-builds the registration information from the module headers. The tests fail right now and doesn't meet PHPCS, so that'll need to be resolved.

To confirm, is the intent of this PR just this module registration and an upcoming PR (or future work on this one?) to build the bits to utilize the registry?

**Comment by enejb** [comment]
Thanks for the review @kraftbj. 

> To confirm, is the intent of this PR just this module registration and an upcoming PR (or future work on this one?) to build the bits to utilize the registry?

Yes more work would be to follow. 
- In future PRs I would like to move more of the code from the main Jetpack class into this Jetpack_Module_registry or other classes where it might make more sense. 
(::activate_new_modules, ::get_default_modules, handle_deprecated_modules, filter_default_modules, translate_module_tag, get_translated_modules, get_active_modules, is_module_active, is_module, activate_default_modules, activate_module, deactivate_module, enable_module_configurable, module_configuration_url, module_configuration_url...)

There is a lot of complexity in the Jetpack class right now that could be removed and we could have tests for as well. 

As well as remove any module information methods from Jetpack_Admin. (`Jetpack_Admin::get_modules()`)

In the future I would also like to remove the current building of php files that are there mostly for translation. (As far as I understand) `tools/build-module-headings-translations.php`.

There could be code in other places that would make sense to remove/move as well.

File: class.jetpack-module.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		$args = wp_parse_args(
+			$args,
+			$default_args
+		);
+		foreach ( $args as $property_name => $property_value ) {
+			$this->$property_name = $property_value;
+		}
+	}
+
+	public function _get_array() {
```
[the remark this answers is not in the document]
Comment (enejb) [reply class.jetpack-module.php]: I am not a big fan of this method and would like to remove it in future PRs. It is there mostly to provide 1:1 mapping with what we currently have.

File: class.jetpack-module.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		 * so please be careful.
+		 *
+		 * @since 3.6.0
+		 *
+		 * @param array   $mod    The details of the requested module.
+		 * @param string  $slug The slug of the module, e.g. sharedaddy
+		 * @param string  $file   The path to the module source file.
+		 */
+		return apply_filters( 'jetpack_get_module', $module_array, $slug, $this->file_path() );
+		return $module_array;
```
[the remark this answers is not in the document]
Comment (enejb) [reply class.jetpack-module.php]: :+1:

File: class.jetpack-modules-registry.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		}
+
+		return self::$instance;
+	}
+
+	public function register( $slug, $args ) {
+		$this->modules[ $slug ] = new Jetpack_Module( $slug, $args );
+	}
+
+	public function get_available_slugs() { // Jetpack::get_available_modules
```
[the remark this answers is not in the document]
Comment (enejb) [reply class.jetpack-modules-registry.php]: I wanted to be more specific in terms of what the method returns. 
Jetpack::get_available_modules currently returns an array of module slugs. 

I am expecting that we get_available_modules would return an array of module objects.

[the remark this answers is not in the document]
Comment (kraftbj) [reply class.jetpack-modules-registry.php]: That makes sense. 👍

File: class.jetpack-module.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	public $free;                       // Free
+	public $requires_connection;        // Requires Connection
+	public $auto_activate;              // Auto Activate
+	public $module_tags;                // Module Tags
+	public $feature;                    // Feature
+	public $additional_search_queries;  // Additional Search Queries
+	public $plan_classes;               // Plans
+
+	public function __construct( $module, $args = array() ) {
+		$this->slug = $module;
```
Comment (lezama) [inline class.jetpack-module.php]: I think if we name `$module` `$slug` it will look prettier?

File: class.jetpack-module.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+		$this->set_props( $args );
+	}
+
+	private function set_props( $args ) {
+		$default_args = array(
```
Comment (lezama) [inline class.jetpack-module.php]: I would set the default when we declare each object property.
i.e:
`public $jumpstart_desc = '';`

File: class.jetpack-modules-registry.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Class for keeping the state of all modules in memory.
+ */
+class Jetpack_Modules_Registry {
+
+	private $modules = array();
+
+	private static $instance = null;
+
+	public static function get_instance() {
```
Comment (lezama) [inline class.jetpack-modules-registry.php]: as there's no constructor maybe the entire class could be `static`? not sure what's the best pattern here

**Comment by lezama** [comment]
Nice work Enej, this has lots of potential, let's keep thinking about this one

File: class.jetpack-module.php (hunk 5 above)
Comment (enejb) [reply class.jetpack-module.php]: I am not sure you can set some defaults as arrays in lower PHP version.

**Comment by dereksmart** [comment]
This is a great chance to move away from the `module` terminology

**Comment by enejb** [comment]
@dereksmart 
Do you know what kind of `terminology` you want to use instread? 
I am a bit very of introducing new terminology, but during this refactor seems like a good idea to do it.

**Comment by enejb** [comment]
@dereksmart Is renaming `module` to `feature` good enough?

**Comment by dereksmart** [comment]
@enejb not sure I should be making that call on my own - it feels like it's worth bikeshedding about in a p2 or something.

**Comment by enejb** [comment]
This is still something that is very much worth exploring. 
Closing it for now. Since we might want to implement it a different way.

**Pull request closed without merging**